File: tests/conformance/test_conformance.m
function tests = test_conformance
%TEST_CONFORMANCE MATLAB route of the cross-language conformance fixture (Task 2.4).
%
%   The permanent parity gate for docs/api-contract-2.0.md §9. Drives the LIVE
%   +dtwc / dtwc_mex pipeline (Problem.set_band -> fill_distance_matrix ->
%   fast_pam -> silhouette/davies_bouldin/dunn) on the recorded conformance
%   dataset and asserts digit-identical labels/medoids and scores within 1e-12
%   rel against conformance_reference.txt (recorded by the C++ route).
%
%   Fixed pipeline (in lockstep with cpp_conformance.cpp, conformance.toml and
%   test_conformance.py): load conformance_series.csv -> banded DTW (band=3) ->
%   FastPAM k=3 -> scores. Labels/medoids are canonicalised (sorted medoid SET;
%   each point labelled by the rank of its assigned medoid) and converted from
%   MATLAB's 1-based MEX boundary to 0-based series indices before comparison.
%
%   Run with:  results = runtests('test_conformance');
%   (Requires the compiled dtwc_mex on the path; otherwise the test is SKIPPED
%    with a loud notice, matching test_contract_parity.m.)
    tests = functiontests(localfunctions);
end

% -------------------------------------------------------------------------
function setupOnce(testCase)
    testCase.TestData.mex_available = (exist('dtwc_mex', 'file') == 3); % 3 == MEX-file
    if ~testCase.TestData.mex_available
        bar = repmat('=', 1, 74);
        warning('dtwc:mexNotBuilt', ['\n' bar '\n' ...
            'SKIPPING dtwc cross-language conformance test.\n' ...
            'Reason: compiled gateway ''dtwc_mex'' not found on the MATLAB path.\n' ...
            'Build with: cmake -B build -DDTWC_BUILD_MATLAB=ON && cmake --build build\n' ...
            'then addpath(build bin) and addpath(''bindings/matlab'').\n' bar]);
    end
    here = fileparts(mfilename('fullpath'));
    testCase.TestData.data_csv  = fullfile(here, 'data', 'conformance_series.csv');
    testCase.TestData.reference = fullfile(here, 'conformance_reference.txt');
    if testCase.TestData.mex_available
        dtwc.device('cpu');
    end
end

function setup(testCase)
    assumeTrue(testCase, testCase.TestData.mex_available, ...
        'dtwc_mex MEX unavailable - skipping (see loud warning above).');
end

% =========================================================================
function test_matlab_route_matches_reference(testCase)
%   Drives dtwc.fast_pam Tier-2 + scores on the recorded conformance dataset.
    N_CLUSTERS = 3;
    BAND = 3;
    MAX_ITER = 100;
    ref = read_reference(testCase.TestData.reference);

    X = readmatrix(testCase.TestData.data_csv);   % 27 x 16, rows = series

    prob = dtwc.Problem('conformance');
    prob.set_data(X);
    prob.set_band(BAND);
    prob.fill_distance_matrix();
    res = dtwc.fast_pam(prob, N_CLUSTERS, 'MaxIter', MAX_ITER); % writes back into prob

    % --- canonicalise (MEX is 1-based; reference is 0-based) ---
    raw_labels  = double(res.labels(:))';           % 1-based index into medoids array
    raw_medoids = double(res.medoid_indices(:))';   % 1-based series indices
    [labels, medoids] = canonicalise(raw_labels, raw_medoids);

    % --- scores (permutation invariant; read prob state written by fast_pam) ---
    sil = dtwc.silhouette(prob);
    silhouette = mean(sil(:));
    davies_bouldin = dtwc.davies_bouldin(prob);
    dunn = dtwc.dunn(prob);

    % Labels + medoids: DIGIT-IDENTICAL.
    verifyEqual(testCase, labels, ref.labels, 'labels differ from reference');
    verifyEqual(testCase, medoids, ref.medoids, 'medoids differ from reference');
    verifyEqual(testCase, numel(medoids), N_CLUSTERS);

    % Scores: equal to 1e-12 relative.
    verify_rel(testCase, silhouette, ref.silhouette, 'silhouette');
    verify_rel(testCase, davies_bouldin, ref.davies_bouldin, 'davies_bouldin');
    verify_rel(testCase, dunn, ref.dunn, 'dunn');
end

% -------------------------------------------------------------------------
function [labels, medoids] = canonicalise(raw_labels, raw_medoids)
%CANONICALISE Sorted 0-based medoid SET + each point labelled by the rank of its
%   assigned medoid. Mirrors cpp_conformance.cpp / test_conformance.py exactly.
    med0 = raw_medoids - 1;                    % 1-based -> 0-based series indices
    medoids = sort(med0);                      % canonical sorted medoid set
    assigned0 = med0(raw_labels);              % 0-based series index of each point's medoid
    labels = zeros(1, numel(assigned0));
    for i = 1:numel(assigned0)
        labels(i) = find(medoids == assigned0(i), 1) - 1;   % 0-based rank
    end
end

function ref = read_reference(path)
    ref = struct();
    lines = readlines(path);
    for i = 1:numel(lines)
        line = strtrim(lines(i));
        if line == "" || startsWith(line, "#"); continue; end
        parts = split(line, ",");
        key = char(parts(1));
        if strcmp(key, 'labels') || strcmp(key, 'medoids')
            ref.(key) = str2double(parts(2:end))';   % 0-based int row vector
        else
            ref.(key) = str2double(parts(2));
        end
    end
end

function verify_rel(testCase, actual, expected, name)
    tol = 1e-12 * max(abs(actual), abs(expected));
    verifyLessThanOrEqual(testCase, abs(actual - expected), tol, ...
        sprintf('%s: %.17g vs reference %.17g exceeds 1e-12 rel', name, actual, expected));
end

File: tests/conformance/test_conformance.py
"""Python + CLI routes of the cross-language conformance fixture (Task 2.4).

The permanent parity gate for docs/api-contract-2.0.md §9. Two live routes assert
against the SAME reference recorded by the C++ route
(``conformance_reference.txt``):

* ``test_python_route_matches_reference`` drives the LIVE ``dtwcpp`` Tier-2
  surface (``Problem.set_band`` -> ``fill_distance_matrix`` -> ``fast_pam`` ->
  ``silhouette``/``davies_bouldin``/``dunn``) — exactly what a Python user calls.
* ``test_cli_route_matches_reference`` invokes the built ``dtwc_cl`` binary with
  the TOML fixture (``conformance.toml``) and asserts its emitted
  ``*_labels.csv`` / ``*_medoids.csv`` / ``*_silhouettes.csv``.

Fixed pipeline (kept in lockstep with cpp_conformance.cpp, conformance.toml and
test_conformance.m): load conformance_series.csv -> banded DTW (band=3) ->
FastPAM k=3 -> scores. Labels/medoids are canonicalised (sorted medoid SET; each
point labelled by the rank of its assigned medoid) so RNG-dependent medoid array
order / cluster-id numbering cannot make identical clusterings compare unequal.
"""
import math
import os
import subprocess
from pathlib import Path

import pytest

import dtwcpp

HERE = Path(__file__).resolve().parent
REPO = HERE.parent.parent
DATA_CSV = HERE / "data" / "conformance_series.csv"
REFERENCE = HERE / "conformance_reference.txt"
TOML = HERE / "conformance.toml"

N_CLUSTERS = 3
BAND = 3
MAX_ITER = 100
SCORE_REL_TOL = 1e-12
# The CLI emits per-point silhouettes at setprecision(8) (dtwc_cl.cpp), so its
# recomputed mean is compared at 1e-6 rel; labels/medoids stay digit-identical.
CLI_SCORE_REL_TOL = 1e-6


# ---------------------------------------------------------------------------
# Shared helpers (mirror cpp_conformance.cpp exactly)
# ---------------------------------------------------------------------------
def load_series(path: Path):
    """Parse the recorded CSV: one comma-separated series per row (no header/id)."""
    rows = []
    for line in path.read_text().splitlines():
        line = line.strip()
        if line:
            rows.append([float(x) for x in line.split(",")])
    return rows


def canonicalise(raw_labels, raw_medoids):
    """Sorted medoid SET + each point labelled by the rank of its assigned medoid.

    ``raw_labels[i]`` indexes into ``raw_medoids`` (FastPAM contract);
    ``raw_medoids[m]`` is a series index.
    """
    sorted_medoids = sorted(int(m) for m in raw_medoids)
    rank = {m: i for i, m in enumerate(sorted_medoids)}
    labels = [rank[int(raw_medoids[int(lab)])] for lab in raw_labels]
    return labels, sorted_medoids


def read_reference(path: Path):
    ref = {}
    for line in path.read_text().splitlines():
        if not line or line.startswith("#"):
            continue
        parts = line.split(",")
        key = parts[0]
        if key in ("labels", "medoids"):
            ref[key] = [int(x) for x in parts[1:]]
        else:
            ref[key] = float(parts[1])
    return ref


def find_cli_binary():
    env = os.environ.get("DTWC_CL_BIN")
    if env and Path(env).exists():
        return Path(env)
    for base in (REPO / "build" / "bin", REPO / "bin", REPO / "build"):
        for name in ("dtwc_cl.exe", "dtwc_cl"):
            cand = base / name
            if cand.exists():
                return cand
    return None


@pytest.fixture(scope="module")
def reference():
    assert REFERENCE.exists(), (
        f"missing {REFERENCE}; regenerate via the C++ route "
        f"(DTWC_CONFORMANCE_REGEN=1 ./bin/cpp_conformance)"
    )
    return read_reference(REFERENCE)


# ---------------------------------------------------------------------------
# Python route — drives the LIVE dtwcpp Tier-2 pipeline
# ---------------------------------------------------------------------------
def test_python_route_matches_reference(reference):
    series = load_series(DATA_CSV)

    prob = dtwcpp.Problem("conformance")
    # names are cosmetic (labels/medoids/scores are name-independent); pass the
    # CLI's 1-based row names so all routes ingest identical Data.
    prob.set_data(series, [str(i + 1) for i in range(len(series))])
    prob.set_band(BAND)
    prob.fill_distance_matrix()
    res = dtwcpp.fast_pam(prob, N_CLUSTERS, MAX_ITER)  # writes labels/medoids back

    labels, medoids = canonicalise(res.labels, res.medoid_indices)

    sil = dtwcpp.silhouette(prob)
    silhouette = sum(sil) / len(sil)
    davies_bouldin = dtwcpp.davies_bouldin(prob)
    dunn = dtwcpp.dunn(prob)

    # Labels + medoids: DIGIT-IDENTICAL.
    assert labels == reference["labels"]
    assert medoids == reference["medoids"]
    assert len(medoids) == N_CLUSTERS

    # Scores: equal to 1e-12 relative.
    assert math.isclose(silhouette, reference["silhouette"], rel_tol=SCORE_REL_TOL)
    assert math.isclose(davies_bouldin, reference["davies_bouldin"], rel_tol=SCORE_REL_TOL)
    assert math.isclose(dunn, reference["dunn"], rel_tol=SCORE_REL_TOL)


# ---------------------------------------------------------------------------
# CLI route — invokes the built dtwc_cl binary with the TOML fixture
# ---------------------------------------------------------------------------
def test_cli_route_matches_reference(reference, tmp_path):
    binary = find_cli_binary()
    if binary is None:
        pytest.skip(
            "=" * 74 + "\nSKIPPING CLI conformance route: dtwc_cl binary not found.\n"
            "Build with: cmake --build build --target dtwc_cl, or set DTWC_CL_BIN.\n" + "=" * 74
        )

    proc = subprocess.run(
        [str(binary), "--config", str(TOML), "-i", str(DATA_CSV), "-o", str(tmp_path)],
        capture_output=True, text=True,
    )
    assert proc.returncode == 0, f"dtwc_cl failed:\n{proc.stdout}\n{proc.stderr}"

    # Parse <name>_labels.csv ("name,cluster") in series-index order.
    labels_rows = (tmp_path / "conformance_labels.csv").read_text().splitlines()[1:]
    raw_labels = [int(r.split(",")[1]) for r in labels_rows if r.strip()]

    # Parse <name>_medoids.csv ("cluster,medoid_index,medoid_name"): row c gives
    # the series index of medoid-array position c (raw_medoids[c]).
    med_rows = (tmp_path / "conformance_medoids.csv").read_text().splitlines()[1:]
    raw_medoids = [0] * N_CLUSTERS
    for r in med_rows:
        if r.strip():
            c, idx = int(r.split(",")[0]), int(r.split(",")[1])
            raw_medoids[c] = idx

    labels, medoids = canonicalise(raw_labels, raw_medoids)

    # Labels + medoids: DIGIT-IDENTICAL to the recorded reference.
    assert labels == reference["labels"]
    assert medoids == reference["medoids"]
    assert len(medoids) == N_CLUSTERS

    # The CLI's own per-point silhouettes -> mean, compared to the reference at the
    # CLI's emitted precision (8 sig figs). Labels/medoids being digit-identical
    # already guarantees every score matches (scores are pure functions of the
    # identical D + partition); this pins the CLI's silhouette output too.
    sil_rows = (tmp_path / "conformance_silhouettes.csv").read_text().splitlines()[1:]
    sil = [float(r.split(",")[2]) for r in sil_rows if r.strip()]
    silhouette = sum(sil) / len(sil)
    assert math.isclose(silhouette, reference["silhouette"], rel_tol=CLI_SCORE_REL_TOL)
